gym_solo/core/obs.py

```diff
@@ -222,7 +258,10 @@ def __init__(self, body_id: int, degrees: bool = False):
     """
     self.robot = body_id
     self._degrees = degrees
-    self._num_joints = p.getNumJoints(self.robot)
+
+  @property
```

Comment:
This is just for my own curiosity. Is it a good practice to declare most/all methods as property?

Reply:
so python is more about the "semantic" meaning behind the attribute/method. Methods are meant to _do_ things while attribute are meant to _be_ things. That's usually why it's a code style to force method names to be verbs. 

In the case of `_num_joints`, it _is_ something rather than _doing_ something; therefore it should be an attribute. However, we know that it depends upon the client to compute the value, which isn't available at instantiation. As per the API requirements, we know that the `client` will be set at registration time, and that `_num_joints` will only be computed in either the `observation_space`, `labels`, or, `compute` functionalities. Therefore, we know that `_num_joints` will only be accessed after the `Observation` has been registered.

Thus, since we need access to the client, we can "fake" this attribute into a method using `@property` so that we can evaluate it lazily. However, its still accessed as `Observation._num_joints` so it's semantically still an attribute. For something like this, I'd use [functool's `@cached_property`](https://docs.python.org/3/library/functools.html#functools.cached_property) for efficiency's sake, but that's unfortunately py3.8+. Since this is just a simple lookup on PyBullet's end, I'm not losing any sleep about the overhead.

Reply:
That's an awesome explanation! Thank you dad!